src/segment-spinner/Segment.ts

```diff
@@ -0,0 +1,229 @@
+      isRequired: state.isRequired,
+      onIncrement: () => state.increment(segment.type),
+      onDecrement: () => state.decrement(segment.type),
+      onIncrementPage: () => state.incrementPage(segment.type),
+      onDecrementPage: () => state.decrementPage(segment.type),
+      onIncrementToMax: () =>
+        state.setSegment(segment.type, segment.maxValue as number),
+      onDecrementToMin: () =>
+        state.setSegment(segment.type, segment.minValue as number),
+    });
+
+    const onKeyDown = (e: any) => {
+      if (e.ctrlKey || e.metaKey || e.shiftKey || e.altKey) {
+        return;
+      }
+
+      switch (e.key) {
+        case "Enter":
+          e.preventDefault();
+          if (segment.isPlaceholder && !state.isReadOnly) {
+            state.confirmPlaceholder(segment.type);
+          }
+          next();
+          break;
+        case "Tab":
+          break;
+        case "Backspace": {
+          e.preventDefault();
+          if (isNumeric(segment.text) && !state.isReadOnly) {
+            const newValue = segment.text.slice(0, -1);
+            state.setSegment(
+              segment.type,
+              newValue.length === 0
+                ? (segment.minValue as number)
+                : parseNumber(newValue),
+            );
+            setEnteredKeys(newValue);
+          }
+          break;
+        }
+        default:
+          e.preventDefault();
+          e.stopPropagation();
+          if ((isNumeric(e.key) || /^[ap]$/.test(e.key)) && !state.isReadOnly) {
+            onInput(e.key);
+          }
+      }
+    };
+
+    const onInput = (key: string) => {
+      const newValue = enteredKeys + key;
+
+      switch (segment.type) {
+        case "dayPeriod":
+          if (key === "a") {
+            state.setSegment("dayPeriod", 0);
+          } else if (key === "p") {
+            state.setSegment("dayPeriod", 12);
+          }
+          next();
+          break;
+        case "day":
+        case "hour":
+        case "minute":
+        case "second":
+        case "month":
+        case "year": {
+          if (!isNumeric(newValue)) {
+            return;
+          }
+
+          const numberValue = parseNumber(newValue);
+          let segmentValue = numberValue;
+          if (
+            segment.type === "hour" &&
+            state.dateFormatter.resolvedOptions().hour12 &&
+            numberValue === 12
+          ) {
+            segmentValue = 0;
+          } else if (numberValue > (segment.maxValue as number)) {
+            segmentValue = parseNumber(key);
+          }
+
+          state.setSegment(segment.type, segmentValue);
+
+          if (Number(numberValue + "0") > (segment.maxValue as number)) {
+            setEnteredKeys("");
+            next();
+          } else {
+            setEnteredKeys(newValue);
+          }
+          break;
+        }
+      }
+    };
+
+    const onFocus = () => {
+      setEnteredKeys("");
+    };
+
+    const id = useId(htmlProps.id);
```

Comment:
Use Reakit's useId

Reply:
Does not seem to be working and causing infinite rerenders. the implementation of both useIds are different.